### PR TITLE
Lazy mobx cache

### PR DESCRIPTION
**Pull request by adamabeshouse**
Implements generic LazyMobXCache class, and uses it to reimplement a few of the mutation table caches.

# Checks
- [x] Follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). For most PRs a single commit should suffice, in some cases multiple topical commits can be useful. During review it is ok to see tiny commits (e.g. Fix reviewer comments), but right before the code gets merged to master or rc branch, any such commits should be squashed since they are useless to the other developers. Definitely avoid [merge commits, use rebase instead.](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [x] Follows the [Airbnb React/JSX Style guide](https://github.com/airbnb/javascript/tree/master/react).
- [x] Make sure your commit messages end with a Signed-off-by string (this line
  can be automatically added by git if you run the `git-commit` command with
  the `-s` option)

### CODE REVIEW

File: src/shared/lib/LazyMobXCache.ts [hunk 1]
```diff
@@ -0,0 +1,171 @@
+import Immutable from "seamless-immutable"; // need to use immutables so mobX can observe the cache shallowly
+import accumulatingDebounce from "./accumulatingDebounce";
+import {observable, action} from "mobx";
+import {AccumulatingDebouncedFunction} from "./accumulatingDebounce";
+
+export type CacheData<T> = {
```
Comment (alisman) [inline src/shared/lib/LazyMobXCache.ts]: we could look into ts namespaces to group these types.

File: src/shared/lib/LazyMobXCache.ts [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        for (const datum of data) {
+            const datumKey = this.dataToKey(datum);
+            toMerge[datumKey] = {
+                status: "complete",
+                data: datum
+            };
+            queryKeyHasData[datumKey] = true;
+        }
+
+        for (const queryKey of Object.keys(queryKeyHasData)) {
```
Comment (alisman) [inline src/shared/lib/LazyMobXCache.ts]: could use a little commenting here to make this easier to understand

File: src/shared/lib/LazyMobXCache.ts (hunk 1 above)
Comment (adamabeshouse) [inline src/shared/lib/LazyMobXCache.ts]: good idea, how do you think it should look?

File: src/shared/lib/LazyMobXCache.ts (hunk 2 above)
Comment (adamabeshouse) [inline src/shared/lib/LazyMobXCache.ts]: done

**Comment by adamabeshouse** [comment]
[generalize caches in a better way, and replace all caches with one cache to rule them all (SampleGeneCache, SimpleCache, CohortVariantCountCache)](https://trello.com/c/sgquyH0w/190-generalize-caches-in-a-better-way-and-replace-all-caches-with-one-cache-to-rule-them-all-samplegenecache-simplecache-cohortvaria)

File: src/shared/lib/LazyMobXCache.ts (hunk 1 above)
Comment (alisman) [inline src/shared/lib/LazyMobXCache.ts]: i'm not totally sure!  I think they are for disambiguating very common names, like Cache for example.  Not a big deal.  I will just merge this and we can revisit namespaces later.

**Pull request merged**